examples/custom_format/main.rs

```diff
@@ -0,0 +1,55 @@
+use std::collections::HashMap;
+
+use config::{Config, File, FileExtensions, Format, Value, ValueKind};
+
+fn main() {
+    let config = Config::builder()
+        .add_source(File::from_str("bad", MyFormat))
+        .add_source(File::from_str("good", MyFormat))
+        .build();
+
+    match config {
+        Ok(cfg) => println!("A config: {:#?}", cfg),
+        Err(e) => println!("An error: {}", e),
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MyFormat;
+
+impl Format for MyFormat {
+    fn parse(
+        &self,
+        uri: Option<&String>,
+        text: &str,
+    ) -> Result<
+        std::collections::HashMap<String, config::Value>,
+        Box<dyn std::error::Error + Send + Sync>,
+    > {
+        // Let's assume our format is somewhat crippled, but this is fine
+        // In real life anything can be used here - nom, serde or other.
+        //
+        // For some more real-life examples refer to format implementation within the library code
+        let mut result = HashMap::new();
+
+        if text == "good" {
+            result.insert(
+                "key".to_string(),
+                Value::new(uri, ValueKind::String(text.into())),
+            );
+        } else {
+            println!("Something went wrong in {:?}", uri);
+        }
+
+        Ok(result)
+    }
+}
+
+// As crazy as it seems for config sourced from a string, legacy demands its sacrifice
+// It is only required for File source, custom sources can use Format without caring for extensions
+static MY_FORMAT_EXT: Vec<&'static str> = vec![];
+impl FileExtensions for MyFormat {
+    fn extensions(&self) -> &Vec<&'static str> {
```

Comment:
So if we have a `&'static str` here, wouldn't it be possible to remove the (heap allocated, if I'm not mistaken) `Vec` here as well?

Maybe with something like a `fn extensions(&self) -> impl Iterator<Item = &'static str>` kind of signature?

Reply:
Good point with the `impl`, more flexible! Static lifetime should be enough, file extensions are quite ... static.

Reply:
To avoid heap allocation, I need to go with `&'static [&'static str]`.

Why?
* `impl Iterator<Item = &'static str>` is forbidden in traits 
* to define associated type Iterator<Item = &'static str> I'd need to return a reference to a slice and I think that to pass lifetime to it I'd need GAT.